libmscore/realizedharmony.cpp

```diff
@@ -0,0 +1,555 @@
+      //bit risky design since these 3 parameters rely on the dirty bit and are not
+      //otherwise checked by RealizedHarmony. This saves us 3 ints of space, but
+      //has the added risk
+      if (!_dirty) {
+            Q_ASSERT(_notes.first() == rootTpc || _notes.first() == bassTpc);
+            return;
+            }
+
+      _notes = generateNotes(rootTpc, bassTpc, _literal, _voicing, transposeOffset);
+      _dirty = false;
+      }
+
+//--------------------------------------------------
+//    getActualDuration
+///    gets the fraction duration for how long
+///    the harmony should be realized based
+///    on the HDuration set.
+///
+///    This is opposed to RealizedHarmony::duration()
+///    which returns the HDuration, which is the duration
+///    setting.
+///
+///    Specifying a parameter finds a duration based on the
+///    passed duration type while not specifying a parameter
+///    uses the setting set by the user for the specific
+///    harmony object.
+//--------------------------------------------------
+Fraction RealizedHarmony::getActualDuration(HDuration durationType) const
+      {
+      HDuration dur;
+      if (durationType != HDuration::INVALID)
+            dur = durationType;
+      else
+            dur = _duration;
+      switch (dur)
+            {
+            case HDuration::UNTIL_NEXT_CHORD_SYMBOL:
+                  return _harmony->ticksTilNext(false);
+                  break;
+            case HDuration::STOP_AT_MEASURE_END:
+                  return _harmony->ticksTilNext(true);
+                  break;
+            case HDuration::SEGMENT_DURATION:
+                  return toSegment(_harmony->parent())->ticks();
+                  break;
+            default:
+                  return Fraction(0, 1);
+            }
+      }
+
+//---------------------------------------------------
+//   getIntervals
+///   gets a map from intervals to TPCs based on
+///   a passed root tpc (this allows for us to
+///   keep pitches, but transpose notes on the score)
+///
+///   Weighting System:
+///   - Rank 0: 3rd, altered 5th, suspensions (characteristic notes)
+///   - Rank 1: 7th
+///   - Rank 2: 9ths
+///   - Rank 3: 13ths where applicable and (in minor chords) 11ths
+///   - Rank 3: Other alterations and additions
+///   - Rank 4: 5th and (in major/dominant chords) 11th
+//---------------------------------------------------
+QMap<int, int> RealizedHarmony::getIntervals(int rootTpc, bool literal) const
+      {
+      //RANKING SYSTEM
+      static const int RANK_MULT = 128;
+      static const int RANK_3RD = 0;
+      static const int RANK_7TH = 1;
+      static const int RANK_9TH = 2;
+      static const int RANK_ADD = 3;
+      static const int RANK_OMIT = 4;
+
+      static const int FIFTH = 7 + RANK_MULT*RANK_OMIT;
+
+      QMap<int, int> ret;
+
+      const ParsedChord* p = _harmony->parsedForm();
+      QString quality = p->quality();
+      int ext = p->extension().toInt();
```

Comment:
One small change coming with my PR #5246: "6/9" and "6,9" will also be single token, so this code might need to special case them.

Reply:
Ah okay that should be manageable. Either that or maybe the extension can be normalized into 69 during parsing.

Reply:
Actually, maybe that already does, see https://github.com/musescore/MuseScore/blob/03c264f057cb4273c9c66936ccb55a7e17d92618/libmscore/chordlist.cpp#L757.

Reply:
Ah yes although looking at the code we might have to have `_extension = 69` in the `if` statement since `_extension` is previously set to `tok1` before all of this extension stuff.